### ISSUE TITLE
StaticArray(T, N)#map returns an Array(T)

### ISSUE THREAD

**Issue by Nephos** [crystal-lang/crystal#3354]
Hello,

I wonder why a `StaticArray.map` returns an `Array` and not a `StaticArray`.

If you think it is better to returns a `StaticArray`, tell me I'll make a PR.

Regards,

**Comment by asterite** [comment]
We could do this, yes. Right now `map` returns `Array` in all types except `Tuple`. There's a bit of tension because `Array` is the recommended container type to use, but I guess optimizing that for `StaticArray` and `Slice` would be a good idea (one can invoke `to_a` at the end if you need an array in the end).

**Comment by Nephos** [comment]
As it is still possible to use ``.map!``, a ``clone.map!`` is equivalent.

I think (correct me if I'm wrong) we cannot change the type of the data stored in a StaticArray in a map
```crystal
    first = yield self[0]
    StaticArray(typeof(first), size).new do |i|
      i == 0 ? first : yield self[i]
    end
```

shouldn't be possible.

Changing ``.map`` behavior is maybe not the right thing to do...

**Comment by Nephos** [comment]
I changed my mind, regarding how ``map`` should behave.
As it is still possible to use ``.map!``, we can also do ``.dup.map!`` (optimizing way).

I think it is not needed to change the behavior of ``StaticArray(T, N).map`` (it means ``map`` cannot return something else than ``StaticArray(T, N)``)

I just added ``.map_with_index!`` which was missing.

- no optimizing problem as it is possible to use ``.dup.map!``
- no non-standard behavior as ``.map`` can returns something else than elements of type T (``StaticArray(T, N)``)

Is it acceptable ?
